File: admin/src/shared/api/agent-runtime.ts
// 作者: 杨永的Agent
// 日期: 2026-05-20
// 修改功能: 场景 **`flowSummary`/`executionSteps`/`closureStatus`** · **`GET …/scenarios/{id}`**（FE_HANDOFF Phase2）
// 作者: 杨永的Agent
// 日期: 2026-05-18
// 修改功能: **`EligibilityEnvelope.requiresMainSite`**（AC-09l · FE_HANDOFF）
// 作者: 杨永的Agent
// 日期: 2026-05-18
// 修改功能: **`IntentRecognizeResponse.effectiveLocale`**（FE_HANDOFF 2026-05-18）
// 作者: 杨永的Agent
// 日期: 2026-05-15
// 修改功能: **GET /scenarios** 响应 **`orchestrationRegistryVersion`**、场景项 **`title`/`category`/`riskLevel`**（FE_HANDOFF · 运行编排页）
// 日期: 2026-05-14
// 修改功能: **现货限价** `POST …/trade/spot/limit-order`（LimitOrderMeta · FE_HANDOFF §2.2）
// 日期: 2026-05-14
// 修改功能: **`RoutingExecuteResponse.executionId`**（落库锚点 · FE_HANDOFF）
// 日期: 2026-05-14
// 修改功能: `routing/execute` Body 可选 **`marketDataLimit`**（depth/trades · FE_HANDOFF）
// 日期: 2026-05-14
// 修改功能: intent/recognize Body 扩展 sessionId 等；Response 对齐 NLU + plan（FE_HANDOFF）
// 日期: 2026-05-18
// 修改功能: **`GET …/trade/spot/open-orders`** · **`POST …/trade/spot/cancel`**（FE_HANDOFF 现货当前委托 / 撤单）
// 日期: 2026-05-13
// 修改功能: Phase2.1 现货闪兑 `GET/POST …/agent/trade/spot/quote|flash-convert` 与类型（FE_HANDOFF）
// 日期: 2026-05-13
// 修改功能: `AgentEvaluateAccessBody` 可选 `telegramChatId`；`capabilities` 补充字段说明
// 日期: 2026-05-12
// 修改功能: Phase1 `/api/v1/agent/*`（含门禁/意图/场景/execution/绑定快照）；routing/execute 增补 `symbol` 与响应 `exchangeReadPreview`

import { HTTPError } from 'ky'

import { AppError } from '@/shared/api/errors'
import { httpClient, toAppError } from '@/shared/api/http-client'

const PREFIX = 'v1/agent'
const SPOT_TRADE_PREFIX = 'v1/agent/trade/spot'

async function normalizeKyError(error: unknown): Promise<never> {
  if (error instanceof HTTPError) {
    const cloned = error.response.clone()
    const body: unknown = await cloned.json().catch(() => null)
    if (
      body !== null &&
      typeof body === 'object' &&
      'message' in body &&
      typeof (body as { message: unknown }).message === 'string'
    ) {
      const code =
        'code' in body && typeof (body as { code: unknown }).code === 'string'
          ? (body as { code: string }).code
          : undefined
      throw new AppError((body as { message: string }).message, {
        status: error.response.status,
        code,
      })
    }
  }
  throw await toAppError(error)
}

/** GET /access/reasons */
export interface AgentAccessReasonItem {
  code: string
  summary: string
}

export interface AgentAccessReasonsResponse {
  reasons: AgentAccessReasonItem[]
}

export async function getAgentAccessReasons(): Promise<AgentAccessReasonsResponse> {
  try {
    return await httpClient.get(`${PREFIX}/access/reasons`).json<AgentAccessReasonsResponse>()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

/** POST /access/evaluate */
export interface AgentEvaluateAccessBody {
  userId: string
  channel?: string
  scenarioId?: string
  /**
   * Telegram 会话 ID（`chat.id`，数值）。
   * 私聊常与 `userId`（tg_id）相同；群 / 超级群 / 话题等不同。
   * Phase1 若按 chat_id 配置环境放行名单，应传此字段以对齐 webhook 门禁。
   */
  telegramChatId?: number
}

export interface AgentEligibilityEnvelope {
  allowed: boolean
  code?: string | null
  reason?: string | null
  locale?: string | null
  /**
   * 扩展能力/诊断字段（JSON 对象，以后端为准）。
   * 示例键：`phase1Allowlist`、`bindingVerified`、`telegramChatId`（可与请求的 `telegramChatId` 对照）。
   */
  capabilities?: Record<string, unknown> | null
  evaluatedAt?: string | null
  eligibilityDecisionId?: string | null
  /**
   * **AC-09l**：是否在「欠托管绑定」语境下提示 **主站 / H5** 引导。
   * - **`true`**：通常 **`allowed=false`** 且 **`AGENT_SUBACCOUNT_REQUIRED`**（未绑定且无放行）。
   * - **`false`**：**`allowed=true`**（放行会话或 **`bindingVerified=true`**）。
   * - **`null` / 字段省略**：封禁 / 灰度 / 全局关闸等其它阻断语义（见 BACKEND_SPEC §3）。
   */
  requiresMainSite?: boolean | null
}

export async function postAgentAccessEvaluate(
  body: AgentEvaluateAccessBody,
): Promise<AgentEligibilityEnvelope> {
  try {
    return await httpClient.post(`${PREFIX}/access/evaluate`, { json: body }).json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

/** POST /intent/recognize */
export interface AgentIntentRecognizeBody {
  text: string
  userId?: string
  channel?: string
  /** 渠道会话键，编排回放 / 可观测（可选） */
  sessionId?: string | null
  /** 当前 execution 锚点（可选） */
  executionId?: string | null
  /** BCP-47 / 产品 locale（可选） */
  locale?: string | null
  /** 上一轮已提交 scenarioId，用于跟进句消歧（可选） */
  previousScenarioId?: string | null
}

export interface AgentIntentCandidate {
  scenarioId: string
  score: number
}

/** NLU：scenarioId 假设条目（keyword_v1） */
export interface AgentIntentScenarioCandidateDraft {
  scenarioId: string
  confidence: number
}

/** NLU 草案（可替换为 LLM 结构化 JSON） */
export interface AgentIntentNluDraft {
  source: 'keyword_v1'
  primaryIntentFamily: string
  scenarioIdCandidates: AgentIntentScenarioCandidateDraft[]
  slots: Record<string, string>
  orderTypeHint: 'market' | 'limit' | 'unknown'
  clarifyHints: string[]
}

export type AgentIntentPlanNextStep =
  | 'CLARIFY'
  | 'ROUTE_READ_SKILL'
  | 'ROUTE_CHAT_FAQ'
  | 'CONFIRM_TYPE_A'
  | 'RESOLVE_FLASH_NOTIONAL'
  | 'RESOLVE_TRADE_NOTIONAL'
  | 'EXECUTE_SPOT_CANCEL'
  | 'EXECUTE_FUTURES_CANCEL'
  | 'BLOCKED_FEATURE'
  | 'STUB_NOT_EXECUTABLE'
  | 'UNKNOWN'

/** 裁决层输出（对接 Bot / Execution 编排） */
export interface AgentIntentPolicyPlan {
  nextStep: AgentIntentPlanNextStep
  resolvedScenarioId?: string | null
  clarify: string[]
  policyCodes: string[]
  note?: string | null
}

export interface AgentIntentRecognizeResponse {
  scenarioId: string | null
  confidence: number
  candidates: AgentIntentCandidate[]
  nlu?: AgentIntentNluDraft | null
  plan?: AgentIntentPolicyPlan | null
  /** 服务端意图路由版本串（可观测 / 回放） */
  orchestrationVersion?: string | null
  /** 与 nlu.source 一致时由服务端回填 */
  nluSource?: string | null
  /** 生效语言：`zh-Hans` | `zh-Hant` | `en` 等（Phase1 FE_HANDOFF 2026-05-18） */
  effectiveLocale?: string | null
}

export async function postAgentIntentRecognize(
  body: AgentIntentRecognizeBody,
): Promise<AgentIntentRecognizeResponse> {
  try {
    return await httpClient.post(`${PREFIX}/intent/recognize`, { json: body }).json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

/** GET /scenarios · Phase2 执行流程登记 */
export interface AgentExecutionFlowStep {
  stepKey: string
  labelZh: string
  order: number
}

export type AgentScenarioReadiness = 'stub' | 'ready' | 'phase1_stub'

export interface AgentScenarioListItem {
  scenarioId: string
  summary: string
  /** 人类可读执行流程（列表优先展示） */
  flowSummary?: string | null
  readiness: AgentScenarioReadiness
  /** 中文短标题（后端寄存器） */
  title?: string | null
  /** read | trade | margin | wealth | automation | chat */
  category?: string | null
  /** low | medium | high */
  riskLevel?: string | null
  /** FROZEN | TBD | PLACEHOLDER */
  closureStatus?: string | null
  executionSteps?: AgentExecutionFlowStep[]
}

export interface AgentScenariosResponse {
  scenarios: AgentScenarioListItem[]
  /** 与控制台寄存器 bundle 对齐，如 `2026.05-orc-v1` */
  orchestrationRegistryVersion?: string | null
}

export interface AgentScenarioDetail {
  scenarioId: string
  summary: string
  flowSummary: string
  readiness: AgentScenarioReadiness
  title?: string | null
  category?: string | null
  riskLevel?: string | null
  closureStatus?: string | null
  flowAnchor?: string | null
  specRefs?: string[]
  promptBindingHint?: string | null
  executionSteps: AgentExecutionFlowStep[]
  orchestrationRegistryVersion?: string | null
}

export async function getAgentScenarios(): Promise<AgentScenariosResponse> {
  try {
    return await httpClient.get(`${PREFIX}/scenarios`).json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

export async function getAgentScenarioDetail(scenarioId: string): Promise<AgentScenarioDetail> {
  const id = scenarioId.trim()
  if (!id) throw new AppError('scenarioId 不可为空', { status: 400 })
  try {
    return await httpClient.get(`${PREFIX}/scenarios/${encodeURIComponent(id)}`).json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

/** POST /routing/execute */
export interface AgentRoutingExecuteBody {
  scenarioId: string
  userId: string
  channel?: string
  utteranceSnapshot?: string
  /** `read.market.ticker` | `read.market.depth` | `read.market.trades` 等只读行情场景由服务端校验 */
  symbol?: string
  /**
   * `read.market.depth`（每侧档位数）/ `read.market.trades`（成交条数上限），整数 **1～100**，默认服务端 **20**。
   */
  marketDataLimit?: number
}

export interface AgentRoutingExecuteResponse {
  routed: boolean
  scenarioId?: string | null
  note?: string | null
  /** 与 Admin `GET …/observability/executions/{executionId}/timeline` 对齐 */
  executionId?: string | null
  /** Phase1：只读所内 JSON 摘要（过滤后），无密钥 */
  exchangeReadPreview?: Record<string, unknown> | null
}

export async function postAgentRoutingExecute(
  body: AgentRoutingExecuteBody,
): Promise<AgentRoutingExecuteResponse> {
  try {
    return await httpClient.post(`${PREFIX}/routing/execute`, { json: body }).json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

/** POST /execution/accept */
export interface AgentExecutionAcceptBody {
  userId: string
  scenarioId?: string
  channel?: string
  idempotencyKey?: string
}

export interface AgentExecutionAcceptResponse {
  executionId: string
  state: string
  createdAt: string
}

export async function postAgentExecutionAccept(
  body: AgentExecutionAcceptBody,
): Promise<AgentExecutionAcceptResponse> {
  try {
    return await httpClient.post(`${PREFIX}/execution/accept`, { json: body }).json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

export interface AgentExecutionStatusResponse {
  executionId: string
  userId: string
  scenarioId?: string | null
  state: string
  createdAt: string
  updatedAt?: string | null
}

export async function getAgentExecution(executionId: string): Promise<AgentExecutionStatusResponse> {
  try {
    return await httpClient.get(`${PREFIX}/execution/${executionId}`).json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

export interface AgentExecutionFinalizeBody {
  executionId: string
  outcome: 'SUCCESS' | 'FAILED' | 'CANCELLED'
  note?: string
}

export interface AgentExecutionFinalizeResponse {
  executionId: string
  state: string
  finalizedAt: string
}

export async function postAgentExecutionFinalize(
  body: AgentExecutionFinalizeBody,
): Promise<AgentExecutionFinalizeResponse> {
  try {
    return await httpClient.post(`${PREFIX}/execution/finalize`, { json: body }).json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

// --- Phase1 onboarding / binding snapshot (§1.1)

export type AgentTradingApiBindingStatus = 'NONE' | 'BOUND'

/** GET /api-binding/status ?userId= */
export interface AgentApiBindingStatusResponse {
  telegramUserId: number
  agentTradingApiBindingStatus: AgentTradingApiBindingStatus
  openapiBaseUrl?: string | null
  bindingId?: number | null
  tgUsername?: string | null
  updatedAt?: string | null
}

export async function getAgentApiBindingStatus(
  userId: string,
): Promise<AgentApiBindingStatusResponse> {
  try {
    return await httpClient
      .get(`${PREFIX}/api-binding/status`, { searchParams: { userId } })
      .json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

/** GET /subaccount/status ?userId= */
export interface AgentSubaccountStatusResponse {
  telegramUserId: number
  subaccountReady: boolean
  agentSubAccountId: string | null
  tradingApiBindingStatus: AgentTradingApiBindingStatus
}

export async function getAgentSubaccountStatus(
  userId: string,
): Promise<AgentSubaccountStatusResponse> {
  try {
    return await httpClient
      .get(`${PREFIX}/subaccount/status`, { searchParams: { userId } })
      .json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

/** POST /onboarding/initiate */
export interface AgentOnboardingInitiateBody {
  telegram?: { tg_id?: string | null }
}

export interface AgentOnboardingInitiateResponse {
  onboardingId: string
  telegramUserId: number
  nextStep: 'bind_trading_api' | 'complete'
  agentTradingApiBindingStatus: AgentTradingApiBindingStatus
}

export async function postAgentOnboardingInitiate(
  body: AgentOnboardingInitiateBody,
): Promise<AgentOnboardingInitiateResponse> {
  try {
    return await httpClient.post(`${PREFIX}/onboarding/initiate`, { json: body }).json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

/** POST /subaccount/create — Phase1 恒为 DEFERRED，非真实创建 */
export interface AgentSubaccountCreatePhase1Response {
  accepted: boolean
  status: 'DEFERRED_EXCHANGE_CONSOLE'
  code: string
  message: string
}

export async function postAgentSubaccountCreate(): Promise<AgentSubaccountCreatePhase1Response> {
  try {
    return await httpClient.post(`${PREFIX}/subaccount/create`, { json: {} }).json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

// --- Phase 2.1 — spot quote + MARKET flash-convert（须托管绑定；见 BACKEND_SPEC §3.3）

/** GET /trade/spot/quote */
export interface AgentTradeSpotQuoteResponse {
  scenarioId: string
  symbol: string
  symbolOrder: string
  /** 公开 ticker 摘要（后端过滤，无密钥） */
  quotePreview: Record<string, unknown>
}

export async function getAgentTradeSpotQuote(
  userId: string,
  symbol: string,
): Promise<AgentTradeSpotQuoteResponse> {
  try {
    return await httpClient
      .get(`${SPOT_TRADE_PREFIX}/quote`, {
        searchParams: { userId, symbol },
      })
      .json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

/** POST /trade/spot/flash-convert */
export interface AgentTradeSpotFlashConvertBody {
  userId: string
  symbol: string
  side: 'BUY' | 'SELL'
  /** 十进制字符串；MARKET BUY 时为计价资产数量（amount） */
  volume: string
  newClientOrderId?: string | null
}

export interface AgentTradeSpotFlashConvertResponse {
  scenarioId: string
  orderId: string | null
  orderIdString: string | null
  clientOrderId: string | null
  status: string | null
  symbol: string | null
  side: string | null
  type: string | null
  exchangeOrderPreview: Record<string, unknown>
}

export async function postAgentTradeSpotFlashConvert(
  body: AgentTradeSpotFlashConvertBody,
): Promise<AgentTradeSpotFlashConvertResponse> {
  try {
    return await httpClient.post(`${SPOT_TRADE_PREFIX}/flash-convert`, { json: body }).json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

/** POST /trade/spot/limit-order · LIMIT，`volume`=base 数量 */
export interface AgentTradeSpotLimitOrderBody {
  userId: string
  symbol: string
  side: 'BUY' | 'SELL'
  /** base 数量十进制字符串 */
  volume: string
  price: string
  timeInForce?: 'GTC' | 'IOC' | 'FOK' | null
  newClientOrderId?: string | null
}

export interface AgentTradeSpotLimitOrderResponse {
  scenarioId: string
  orderId: string | null
  orderIdString: string | null
  clientOrderId: string | null
  status: string | null
  symbol: string | null
  side: string | null
  type: string | null
  exchangeOrderPreview: Record<string, unknown>
}

export async function postAgentTradeSpotLimitOrder(
  body: AgentTradeSpotLimitOrderBody,
): Promise<AgentTradeSpotLimitOrderResponse> {
  try {
    return await httpClient.post(`${SPOT_TRADE_PREFIX}/limit-order`, { json: body }).json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

/** GET /trade/spot/open-orders · Coobit GET /sapi/v2/openOrders */
export interface AgentTradeSpotOpenOrdersResponse {
  scenarioId: string
  symbol?: string | null
  symbolOrder?: string | null
  orders: Record<string, unknown>[]
}

export async function getAgentTradeSpotOpenOrders(
  userId: string,
  opts?: { symbol?: string; limit?: number },
): Promise<AgentTradeSpotOpenOrdersResponse> {
  const searchParams: Record<string, string | number> = { userId }
  const sym = opts?.symbol?.trim()
  if (sym) searchParams.symbol = sym
  if (opts?.limit != null && Number.isFinite(opts.limit)) {
    searchParams.limit = opts.limit
  }
  try {
    return await httpClient
      .get(`${SPOT_TRADE_PREFIX}/open-orders`, { searchParams })
      .json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

/** POST /trade/spot/cancel · Coobit POST /sapi/v2/cancel */
export interface AgentTradeSpotCancelBody {
  userId: string
  symbol: string
  orderId?: string | null
  newClientOrderId?: string | null
}

export interface AgentTradeSpotCancelResponse {
  scenarioId: string
  orderId?: string | null
  orderIdString?: string | null
  clientOrderId?: string | null
  status?: string | null
  symbol?: string | null
  side?: string | null
  type?: string | null
  exchangeOrderPreview: Record<string, unknown>
}

export async function postAgentTradeSpotCancel(
  body: AgentTradeSpotCancelBody,
): Promise<AgentTradeSpotCancelResponse> {
  try {
    return await httpClient.post(`${SPOT_TRADE_PREFIX}/cancel`, { json: body }).json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

File: server/chainup_agent/api/schemas/agent_runtime.py
"""Runtime Phase 1 DTOs (access / intent / routing / execution) aligned with OpenAPI fragments."""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Any, Literal, Self

from pydantic import BaseModel, ConfigDict, Field, model_validator

# --- Intent pipeline (NLU draft + policy plan) ---


class IntentScenarioCandidateDraft(BaseModel):
    """Structured NLU: one scenarioId hypothesis with score."""

    model_config = ConfigDict(populate_by_name=True)

    scenario_id: str = Field(alias="scenarioId")
    confidence: float = Field(ge=0.0, le=1.0)


class IntentNluDraft(BaseModel):
    """Understanding-layer output (keyword MVP; replaceable by LLM structured JSON)."""

    model_config = ConfigDict(populate_by_name=True)

    source: Literal["keyword_v1", "llm_structured_v1"] = Field(
        description="NLU backend: keyword_v1 | llm_structured_v1 (LLM JSON with keyword merge).",
    )
    primary_intent_family: str = Field(
        alias="primaryIntentFamily",
        description="Coarse bucket: trade_write | portfolio_read | wealth_read | chat | unknown",
    )
    scenario_id_candidates: list[IntentScenarioCandidateDraft] = Field(
        default_factory=list,
        alias="scenarioIdCandidates",
    )
    slots: dict[str, str] = Field(default_factory=dict)
    order_type_hint: Literal["market", "limit", "unknown"] = Field(
        default="unknown",
        alias="orderTypeHint",
    )
    clarify_hints: list[str] = Field(default_factory=list, alias="clarifyHints")


class IntentPolicyPlan(BaseModel):
    """Deterministic router output — feeds execution / Telegram orchestration."""

    model_config = ConfigDict(populate_by_name=True)

    next_step: Literal[
        "CLARIFY",
        "ROUTE_READ_SKILL",
        "ROUTE_CHAT_FAQ",
        "CONFIRM_TYPE_A",
        "RESOLVE_FLASH_NOTIONAL",
        "RESOLVE_TRADE_NOTIONAL",
        "EXECUTE_SPOT_CANCEL",
        "EXECUTE_FUTURES_CANCEL",
        "BLOCKED_FEATURE",
        "STUB_NOT_EXECUTABLE",
        "UNKNOWN",
    ] = Field(alias="nextStep")
    resolved_scenario_id: str | None = Field(default=None, alias="resolvedScenarioId")
    clarify: list[str] = Field(default_factory=list)
    policy_codes: list[str] = Field(default_factory=list, alias="policyCodes")
    note: str | None = None


# --- Access (access-control EvaluateEligibility + EligibilityEnvelope) ---


class EvaluateEligibilityRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    user_id: str | None = Field(
        default=None,
        alias="userId",
        description=(
            "Telegram user id (tg_id) as numeric string — NOT the exchange sub-account id. "
            "Omit if subAccountId is provided."
        ),
    )
    sub_account_id: str | None = Field(
        default=None,
        alias="subAccountId",
        description=(
            "Exchange sub-account id from binding confirm; server resolves Telegram tg_id via agent_instance."
        ),
    )
    channel: str | None = Field(
        default=None,
        description="e.g. telegram; default treated as telegram in Coobit",
    )
    scenario_id: str | None = Field(default=None, alias="scenarioId")
    telegram_chat_id: int | None = Field(
        default=None,
        alias="telegramChatId",
        description=(
            "Optional Telegram chat.id from webhook envelope; "
            "allowlist (CHAINUP_AGENT_TELEGRAM_BOUND_CHAT_ALLOWLIST) may match chat id"
        ),
    )

    @model_validator(mode="after")
    def _user_or_subaccount(self) -> Self:
        u = (self.user_id or "").strip()
        s = (self.sub_account_id or "").strip()
        if not u and not s:
            raise ValueError("须提供 userId（Telegram tg_id）或 subAccountId（子账户 ID）至少一项。")
        return self


class EligibilityEnvelope(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    allowed: bool
    code: str | None = None
    reason: str | None = None
    locale: str | None = None
    capabilities: dict[str, Any] | None = None
    requires_main_site: bool | None = Field(
        default=None,
        alias="requiresMainSite",
        description=(
            "conservative hint: true when UX should route to Deeplink/H5/main-site binding "
            "or onboarding (e.g. subaccount/bind required); false when Telegram-only session is sufficient."
        ),
    )
    evaluated_at: datetime | None = Field(default=None, alias="evaluatedAt")
    eligibility_decision_id: str | None = Field(default=None, alias="eligibilityDecisionId")


class AccessReasonItem(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    code: str
    summary: str


class AccessReasonsResponse(BaseModel):
    reasons: list[AccessReasonItem]


# --- Intent ---


class IntentRecognizeRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    text: str = Field(description="User-visible utterance channel text")
    user_id: str | None = Field(default=None, alias="userId")
    channel: str | None = None
    session_id: str | None = Field(
        default=None,
        alias="sessionId",
        description="Channel session key for replay / observability (optional Phase1).",
    )
    execution_id: str | None = Field(
        default=None,
        alias="executionId",
        description="Active execution anchor if any (optional).",
    )
    locale: str | None = Field(default=None, description="BCP-47 / product locale hint")
    previous_scenario_id: str | None = Field(
        default=None,
        alias="previousScenarioId",
        description="Last committed scenarioId in the session (follow-up disambiguation).",
    )


class IntentCandidate(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    scenario_id: str = Field(alias="scenarioId")
    score: float = Field(ge=0.0, le=1.0)


class IntentRecognizeResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    scenario_id: str | None = Field(default=None, alias="scenarioId")
    confidence: float = Field(ge=0.0, le=1.0)
    candidates: list[IntentCandidate] = Field(default_factory=list)
    nlu: IntentNluDraft | None = None
    plan: IntentPolicyPlan | None = None
    orchestration_version: str | None = Field(
        default=None,
        alias="orchestrationVersion",
        description="Server intent-router version string (observability / replay).",
    )
    nlu_source: str | None = Field(
        default=None,
        alias="nluSource",
        description="Same as nlu.source when present.",
    )
    effective_locale: str | None = Field(
        default=None,
        alias="effectiveLocale",
        description="Normalized locale bucket: zh-Hans | zh-Hant | en (from request locale hint).",
    )
    effective_intent_nlu_use_llm: bool | None = Field(
        default=None,
        alias="effectiveIntentNluUseLlm",
        description=(
            "Whether this request entered the LLM NLU attempt branch "
            "(env override or gateway_defaults.intentNluUseLlm)."
        ),
    )


# --- Routing (noop body) ---


class RoutingExecuteRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    scenario_id: str = Field(alias="scenarioId")
    user_id: str = Field(alias="userId")
    channel: str | None = None
    utterance_snapshot: str | None = Field(
        default=None,
        alias="utteranceSnapshot",
        description="Optional echo of triggering text for audits",
    )
    symbol: str | None = Field(
        default=None,
        description=(
            "read.market.* public market: trading pair (e.g. BTC-USDT), "
            "normalized server-side for /sapi/v2/ticker|depth|trades"
        ),
    )
    market_data_limit: int | None = Field(
        default=None,
        alias="marketDataLimit",
        ge=1,
        le=100,
        description=(
            "read.market.depth (per-side rows) and read.market.trades (trade count); "
            "default 20 when omitted."
        ),
    )


class RoutingExecuteResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    routed: bool
    scenario_id: str | None = Field(default=None, alias="scenarioId")
    note: str | None = None
    exchange_read_preview: dict[str, Any] | None = Field(
        default=None,
        alias="exchangeReadPreview",
        description="read.* exchange JSON snapshot (filtered); no secrets.",
    )
    execution_id: str | None = Field(
        default=None,
        alias="executionId",
        description=(
            "Set when POST /routing/execute runs the HTTP execution lifecycle "
            "(accept → exchange → finalize); use for Admin timeline correlation."
        ),
    )


# --- Scenarios catalogue ---


class ExecutionFlowStepOut(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    step_key: str = Field(alias="stepKey")
    label_zh: str = Field(alias="labelZh")
    order: int


class ScenarioListItem(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    scenario_id: str = Field(alias="scenarioId")
    summary: str
    flow_summary: str | None = Field(
        default=None,
        alias="flowSummary",
        description="Human-readable 执行流程（与 summary 同窗，Phase2 优先展示本字段）。",
    )
    readiness: Literal["stub", "ready"]
    title: str | None = Field(
        default=None,
        description="Short zh title — aligned with Admin 运行场景 registry.",
    )
    category: str | None = Field(
        default=None,
        description="read | trade | margin | wealth | automation | chat",
    )
    risk_level: str | None = Field(
        default=None,
        alias="riskLevel",
        description="low | medium | high (product risk hint for ops UI).",
    )
    closure_status: str | None = Field(
        default=None,
        alias="closureStatus",
        description="FROZEN | TBD | PLACEHOLDER — ops runtime status hint.",
    )
    execution_steps: list[ExecutionFlowStepOut] = Field(
        default_factory=list,
        alias="executionSteps",
        description="Registered orchestration step sequence (FR-AO01).",
    )


class ScenarioDetailOut(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    scenario_id: str = Field(alias="scenarioId")
    title: str | None = None
    category: str | None = None
    risk_level: str | None = Field(default=None, alias="riskLevel")
    readiness: Literal["stub", "ready"]
    closure_status: str | None = Field(default=None, alias="closureStatus")
    flow_summary: str = Field(alias="flowSummary")
    summary: str
    flow_anchor: str | None = Field(default=None, alias="flowAnchor")
    spec_refs: list[str] = Field(default_factory=list, alias="specRefs")
    prompt_binding_hint: str | None = Field(default=None, alias="promptBindingHint")
    execution_steps: list[ExecutionFlowStepOut] = Field(
        default_factory=list,
        alias="executionSteps",
    )
    orchestration_registry_version: str | None = Field(
        default=None,
        alias="orchestrationRegistryVersion",
    )


class ScenarioListResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    scenarios: list[ScenarioListItem]
    orchestration_registry_version: str | None = Field(
        default=None,
        alias="orchestrationRegistryVersion",
        description="Registry bundle id — same spirit as Admin ORCHESTRATION_VERSION_DISPLAY.",
    )


# --- Execution (persisted ``agent_execution``; billing / audit anchor) ---


class ExecutionAcceptRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    user_id: str = Field(alias="userId")
    scenario_id: str | None = Field(default=None, alias="scenarioId")
    channel: str | None = None
    idempotency_key: str | None = Field(default=None, alias="idempotencyKey")
    prompt_pack_version: str | None = Field(
        default=None,
        alias="promptPackVersion",
        description="Optional; when omitted, server fills from published pack for scenarioId if any.",
    )
    resolved_prompt_binding: dict[str, Any] | None = Field(
        default=None,
        alias="resolvedPromptBinding",
        description="Optional resolved prompt binding snapshot (observability).",
    )


class ExecutionAcceptResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    execution_id: str = Field(alias="executionId")
    state: Literal["ACCEPTED"] = "ACCEPTED"
    created_at: datetime = Field(alias="createdAt")


class ExecutionStatusResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    execution_id: str = Field(alias="executionId")
    user_id: str = Field(alias="userId")
    scenario_id: str | None = Field(default=None, alias="scenarioId")
    state: str
    created_at: datetime = Field(alias="createdAt")
    updated_at: datetime | None = Field(default=None, alias="updatedAt")
    prompt_pack_version: str | None = Field(
        default=None,
        alias="promptPackVersion",
        description="Published prompt pack version when scenario had an effective pack.",
    )
    resolved_prompt_binding: dict[str, Any] | None = Field(
        default=None,
        alias="resolvedPromptBinding",
    )


class ExecutionFinalizeRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    execution_id: str = Field(alias="executionId")
    outcome: Literal["SUCCESS", "FAILED", "CANCELLED", "UNKNOWN"]
    note: str | None = None


class ExecutionFinalizeResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    execution_id: str = Field(alias="executionId")
    state: str
    finalized_at: datetime = Field(alias="finalizedAt")


def utc_now() -> datetime:
    return datetime.now(UTC)
